File: gitdot-core/src/dto/domain/core/review/review_review_diff.rs
use uuid::Uuid;

use crate::{
    dto::common::{OwnerName, RepositoryName},
    error::{InputError, ReviewError},
    model::CommentSide,
};

#[derive(Debug, Clone)]
pub struct ReviewCommentInput {
    pub revision_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub body: String,
    pub file_path: Option<String>,
    pub line_number_start: Option<i32>,
    pub line_number_end: Option<i32>,
    pub start_character: Option<i32>,
    pub end_character: Option<i32>,
    pub side: Option<CommentSide>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ReviewAction {
    Comment,
    Approve,
    Reject,
}

#[derive(Debug, Clone)]
pub struct ReviewReviewDiffRequest {
    pub owner: OwnerName,
    pub repo: RepositoryName,
    pub number: i32,
    pub position: i32,
    pub reviewer_id: Uuid,
    pub action: ReviewAction,
    pub comments: Vec<ReviewCommentInput>,
}

impl ReviewReviewDiffRequest {
    pub fn new(
        owner: &str,
        repo: &str,
        number: i32,
        position: i32,
        reviewer_id: Uuid,
        action: &str,
        comments: Vec<(
            Uuid,
            Option<Uuid>,
            String,
            Option<String>,
            Option<i32>,
            Option<i32>,
            Option<i32>,
            Option<i32>,
            Option<String>,
        )>,
    ) -> Result<Self, ReviewError> {
        let action = match action {
            "comment" => ReviewAction::Comment,
            "approve" => ReviewAction::Approve,
            "reject" => ReviewAction::Reject,
            other => {
                return Err(InputError::new(
                    "action",
                    format!("Invalid action: {other}. Must be comment, approve, or reject"),
                )
                .into());
            }
        };

        let comments = comments
            .into_iter()
            .map(
                |(
                    revision_id,
                    parent_id,
                    body,
                    file_path,
                    line_number_start,
                    line_number_end,
                    start_character,
                    end_character,
                    side,
                )| {
                    let side = side
                        .as_deref()
                        .map(|s| -> Result<CommentSide, ReviewError> {
                            match s {
                                "old" => Ok(CommentSide::Old),
                                "new" => Ok(CommentSide::New),
                                _ => Err(InputError::new(
                                    "side",
                                    format!("Invalid side: {s}. Must be old or new"),
                                )
                                .into()),
                            }
                        })
                        .transpose()?;

                    Ok(ReviewCommentInput {
                        revision_id,
                        parent_id,
                        body,
                        file_path,
                        line_number_start,
                        line_number_end,
                        start_character,
                        end_character,
                        side,
                    })
                },
            )
            .collect::<Result<Vec<_>, ReviewError>>()?;

        Ok(Self {
            owner: OwnerName::try_new(owner).map_err(|e| InputError::new("owner name", e))?,
            repo: RepositoryName::try_new(repo)
                .map_err(|e| InputError::new("repository name", e))?,
            number,
            position,
            reviewer_id,
            action,
            comments,
        })
    }
}

File: gitdot-web/packages/gitdot-api/src/endpoint/review/review-review-diff.ts
import { z } from "zod";
import { ReviewResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ReviewCommentInput = z.object({
  revision_id: z.string().uuid(),
  parent_id: z.string().uuid().optional(),
  body: z.string(),
  file_path: z.string().optional(),
  line_number_start: z.number().int().optional(),
  line_number_end: z.number().int().optional(),
  start_character: z.number().int().optional(),
  end_character: z.number().int().optional(),
  side: z.string().optional(),
});
export type ReviewCommentInput = z.infer<typeof ReviewCommentInput>;

export const ReviewReviewDiffRequest = z.object({
  action: z.enum(["comment", "approve", "reject"]),
  comments: z.array(ReviewCommentInput),
});
export type ReviewReviewDiffRequest = z.infer<typeof ReviewReviewDiffRequest>;

export const ReviewReviewDiffResponse = ReviewResource;
export type ReviewReviewDiffResponse = z.infer<typeof ReviewReviewDiffResponse>;

export const ReviewReviewDiff = {
  path: "/repository/{owner}/{repo}/review/{number}/diff/{position}/review",
  method: "POST",
  request: ReviewReviewDiffRequest,
  response: ReviewReviewDiffResponse,
} as const satisfies Endpoint;
export type ReviewReviewDiff = typeof ReviewReviewDiff;
